### ISSUE TITLE
Test suite: mam4_jax import leaks jax_enable_x64=True process-wide; full single-process run segfaults on Derecho

### ISSUE THREAD

**Issue by duncanwp** [climate-analytics-lab/jax-gcm#729]
Two coupled test-infrastructure problems found while trying to reproduce the CI gate locally on Derecho (2026-08-25 session, branch `fix/702-emulator-radiation`):

## 1. `import mam4_jax` sets `jax_enable_x64=True` process-wide

Verified directly: `jax.config.jax_enable_x64` flips `False → True` across the import. When the `jcm[mam4]` extra is installed and tests run under xdist (`-n 12`), any worker that imports a mam4-touching test first silently runs every later test on that worker in float64 — producing ~15 reproducible, worker-placement-dependent failures that look like real numerical regressions (disjoint failure sets between branches, not flakes; confirmed by re-run).

CI never sees this because it installs `pip install -e .` with **no extras** and runs single-process (`run_test.yaml:92,98`) — but anyone measuring "CI-equivalent" results locally with the extra installed gets garbage.

Fix directions: the root `conftest.py` (which already clears `jcm` imports between tests) could snapshot/restore `jax_enable_x64`; and/or upstream a fix to mam4-jax so it scopes its precision requirement instead of mutating global config at import.

## 2. Full suite single-process segfaults on Derecho login nodes

The exact CI command (single-process, with coverage) segfaults after ~20 min on `derecho2`; a targeted `jcm/physics/radiation/` run also segfaults at ~85% with no failing test. Login-node quirks are known (`nproc` reports 1 while `sched_getaffinity` shows 128; big suites dump core), and the practical workaround is running via PBS `-q develop` — but the segfault has not been reproduced/excluded on a compute node with the CI's exact dependency set, so the "CI-equivalent gate passes locally" claim currently rests on inference, not measurement.

Deliverable: a documented, reproducible local CI-equivalent recipe for Derecho (probably in the `jcm-local-ci` skill + `docs/source/developer.rst`): PBS develop node, no optional extras, single-process or xdist-with-x64-guard.

**Comment by duncanwp** [comment]
Root cause narrowed (same-day follow-up). There are **two** process-wide x64 flippers, and the import-time one in mam4-jax is *not* the operative leak in this repo's suite:

1. **`jcm`'s own adapter**: `Mam4JaxAdapter.__init__` (`jcm/physics/aerosol/jam/microphysics/mam4_jax.py:224`) calls `jax.config.update("jax_enable_x64", want_x64)` at construction, with `want_x64` defaulting to True via `MAM4_JAX_ENABLE_X64` (not `JAX_ENABLE_X64` — mam4-jax's own import-time opt-out does not reach this call). Any worker that runs `mam4_jax_test.py` flips itself for every subsequent test.
2. **pySES** (`FVPhysicsGrid`/`PysesCamSEDycore` construction) — already mitigated by `jcm/dycore/pyses/conftest.py`'s collection reordering, **but `pytest-randomly` re-shuffles after that hook**, defeating it. Any local environment with pytest-randomly installed must pass `-p no:randomly` for the reordering to hold.

Measured on a Derecho develop node (branch `fix/702-emulator-radiation`, `-n 8 -m "not slow"`):
- plain: 15 dtype-sensitive failures; on `origin/dev` baseline: 5 *different* dtype-sensitive failures (disjoint sets — worker-placement-dependent, exactly as previously observed).
- `JAX_ENABLE_X64=0` + per-test restore fixture: unchanged (confirms the adapter call, not the import, is the leak).
- All failures pass serially.

**CI-equivalent local recipe** (CI installs no extras, so `mam4_jax_test.py` skips there):

```bash
JAX_PLATFORMS=cpu python -m pytest -n 8 -m "not slow" -q -p no:randomly     --ignore=jcm/physics/aerosol/jam/microphysics/mam4_jax_test.py
```

Candidate real fixes: (a) the adapter stops mutating global config and instead asserts/uses a scoped dtype context (its `core_dtype='float32'` path already works with x64 off); (b) a conftest reordering hook for `mam4_jax_test.py` like the pySES one; (c) document `MAM4_JAX_ENABLE_X64=0` for local xdist runs (but note some adapter tests may then legitimately diverge from their x64 expectations).

**Comment by duncanwp** [comment]
Negative result worth recording: an instrumented develop-node run (`-n 8 -p no:randomly`, mam4 test ignored, autouse fixture logging any `jax_enable_x64` change across each test body, verified twice) reproduced 13 dtype-flavoured failures with **zero flips recorded**. Combined with serial runs passing (which perform the same collection-time imports), this rules out `jax_enable_x64` leakage — import-time *and* test-time — as the mechanism for the phantom xdist failures. The adapter/pySES flippers described above are real and still worth fixing, but they are not what breaks these runs.

Next step for whoever picks this up: capture full tracebacks from an xdist run (the PBS logs here kept only tails) and check other process-global leak candidates — `jcm.constants.set_constants` callers, `jax_default_matmul_precision`, rank promotion, or compilation-cache interactions. The failing set varies run-to-run at fixed code, so it is scheduling-dependent shared state, not code.
